turbo/snapshotsync/snapshotsync.go

```diff
@@ -66,11 +70,182 @@ func RequestSnapshotsDownload(ctx context.Context, downloadRequest []services.Do
 	return nil
 }
 
+const (
+	MinPruneStep = 64
+)
+
+func adjustStepPrune(steps uint) uint {
+	if steps < MinPruneStep {
+		return MinPruneStep
+	}
+	if steps%MinPruneStep == 0 {
+		return steps
+	}
+	// round to nearest multiple of 64. if less than 64, round to 64
+	return steps + steps%MinPruneStep
+}
+
+func adjustBlockPrune(blocks, minBlocksToDownload uint) uint {
+	if minBlocksToDownload < snaptype.Erigon2MergeLimit {
+		minBlocksToDownload = snaptype.Erigon2MergeLimit
+	}
+	if blocks < minBlocksToDownload {
+		return minBlocksToDownload
+	}
+	if blocks%snaptype.Erigon2MergeLimit == 0 {
+		return blocks
+	}
+	ret := blocks + snaptype.Erigon2MergeLimit
+	// round to nearest multiple of 64. if less than 64, round to 64
+	return ret - ret%snaptype.Erigon2MergeLimit
+}
+
+func shouldUseStepsForPruning(name string) bool {
+	return strings.HasPrefix(name, "idx") || strings.HasPrefix(name, "history")
+}
+
+func canSnapshotBePruned(name string) bool {
+	return strings.HasPrefix(name, "idx") || strings.HasPrefix(name, "history") || strings.Contains(name, "transactions")
+}
+
+func buildBlackListForPruning(pruneMode bool, stepPrune, minBlockToDownload, blockPrune uint, preverified snapcfg.Preverified) (map[string]struct{}, error) {
+	type snapshotFileData struct {
+		from, to  uint64
+		stepBased bool
+		name      string
+	}
+	blackList := make(map[string]struct{})
+	if !pruneMode {
+		return blackList, nil
+	}
+	stepPrune = adjustStepPrune(stepPrune)
+	blockPrune = adjustBlockPrune(blockPrune, minBlockToDownload)
+	snapshotKindToNames := make(map[string][]snapshotFileData)
+	for _, p := range preverified {
+		name := p.Name
+		// Dont prune unprunable files
+		if !canSnapshotBePruned(name) {
+			continue
+		}
+		var from, to uint64
+		var err error
+		var kind string
+		if shouldUseStepsForPruning(name) {
+			// parse "from" (0) and "to" (64) from the name
+			// parse the snapshot "kind". e.g kind of 'idx/v1-accounts.0-64.ef' is "idx/v1-accounts"
+			rangeString := strings.Split(name, ".")[1]
+			rangeNums := strings.Split(rangeString, "-")
+			// convert the range to uint64
+			from, err = strconv.ParseUint(rangeNums[0], 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			to, err = strconv.ParseUint(rangeNums[1], 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			kind = strings.Split(name, ".")[0]
+		} else {
+			// e.g 'v1-000000-000100-beaconblocks.seg'
+			// parse "from" (000000) and "to" (000100) from the name. 100 is 100'000 blocks
+			minusSplit := strings.Split(name, "-")
+			s, _, _ := snaptype.ParseFileName("tmp", name)
```

Comment:
check `ok`

Reply:
ok returns whether it is a state file or not...

```
func ParseFileName(dir, fileName string) (res FileInfo, isE3Seedable bool, ok bool) {
	res, ok = parseFileName(dir, fileName)
	if ok {
		return res, false, true
	}
	isStateFile := IsStateFile(fileName)
	res.name = fileName
	res.Path = filepath.Join(dir, fileName)
	return res, isStateFile, isStateFile
}
```
in that branch it is never a state file. do you want me to sanity check and if ok == true to skip it?

Reply:
still - check `ok` it's like check `err` - most of the time they do not happen, but sometime code moving or condition changing. and if don't check `ok` then `nil-ptr` may happen. linters must catch unhandled `ok` same as they catching unhandled `err`.